### ISSUE TITLE
[Training] Add opt-in torch.compile acceleration with measured per-model support

### ISSUE THREAD

**Issue by Ikkjo** [Ikkjo/neural-fx#40]
## Context

Training currently uses eager PyTorch. The project is pinned to PyTorch 2.11 and trains LSTM, GRU, WaveNet, and S4D through PyTorch Lightning. Current GPU profiling suggests the convolutional/state-space models are the most promising compilation candidates, but support and benefit must be measured per architecture:

- built-in LSTM/GRU execution may graph-break or remain mostly eager;
- the loss fallback currently calls `loss.item()`, which introduces a host synchronization and a data-dependent graph break;
- recurrent/streaming state mutation may trigger graph breaks or recompilation;
- a compiled wrapper must not leak into checkpoint keys or change ONNX/TorchScript export behavior.

This issue adds a controlled training optimization, not a blanket `torch.compile(...)` call.

## Goal

Add opt-in `torch.compile` support to the standard training path, validate correctness and checkpoint/export compatibility, and publish reproducible eager-versus-compiled GPU results for every model family. Keep eager training as the default until the measurements justify changing a model's recommended configuration.

## Proposed interface

Add a small typed compile section to the training config and equivalent CLI overrides:

```yaml
training:
  compile:
    enabled: false
    mode: default
    cache_dir: null
```

Support `default`, `reduce-overhead`, and `max-autotune` modes only if they work with the pinned PyTorch version. Record the requested and effective compiler settings in run metadata. An explicitly requested compile failure must be actionable; do not silently continue in eager mode.

Compile only the training execution path. Checkpoints must retain the canonical eager model state-dict schema, and export must reconstruct a fresh eager model before producing ONNX, TorchScript, or RTNeural artifacts.

When compilation is enabled, `cache_dir: null` resolves to a persistent user cache outside the repository. Use `$XDG_CACHE_HOME/neural-fx/torchinductor` when available and `~/.cache/neural-fx/torchinductor` otherwise. An existing `TORCHINDUCTOR_CACHE_DIR` or explicit config/CLI value takes precedence. Print the resolved cache directory at startup.

### 1. Establish an eager baseline

- [ ] Add a reproducible benchmark command or extend the existing benchmark tooling to measure training, not inference.
- [ ] Fix the workload: real paired audio, fixed model/config/seed, batch size, segment length, precision, optimizer, warm-up, and measured steps.
- [ ] Record cold startup/first-step time separately from warmed optimizer-step or epoch time, throughput, peak CUDA memory, device/runtime versions, and final validation loss.
- [ ] Capture at least three fresh-process eager runs for LSTM, GRU, WaveNet, and S4D.

### 2. Add the compile integration seam

- [ ] Add typed config parsing, YAML defaults, and CLI overrides; compilation is off by default.
- [ ] Apply compilation at one documented boundary compatible with Lightning, optimizer creation, callbacks, and distributed/device placement.
- [ ] Preserve stable parameter names and checkpoint keys; avoid persisting an `OptimizedModule`/`_orig_mod` wrapper.
- [ ] Log the effective compiler mode and whether compilation was actually active.
- [ ] Keep input shapes stable for the configured batch/segment workload where possible.

### 3. Add persistent local caching and a fast cache-prime path

- [ ] Resolve a persistent per-user cache directory outside the repository whenever compilation is enabled.
- [ ] Add `cache_dir` config and `--compile-cache-dir` CLI overrides while honoring an existing `TORCHINDUCTOR_CACHE_DIR`.
- [ ] Enable the local FX graph and AOTAutograd caches for compiled training unless the user explicitly disables them. Let Triton use the same Inductor cache tree.
- [ ] Create the cache directory on demand and print it with the effective compile settings.
- [ ] Add a dedicated command:

  ```bash
  python scripts/prime_compile_cache.py \
      --config configs/models/wavenet/config.yaml \
      --device cuda
  ```

- [ ] Make the prime command reuse the production model, loss, optimizer, and compile construction code so it cannot drift from `scripts/train.py`.
- [ ] Derive batch size, segment length, dtype, TBPTT behavior, model architecture, and compile mode from the supplied training config.
- [ ] Use seeded synthetic input and target tensors with the exact configured shapes. Cache keys depend on the graph and tensor metadata, so cache priming must not read or process the full audio dataset.
- [ ] Run enough work to trigger lazy compilation and autotuning for the training forward pass, loss, backward pass, and any compiled optimizer work. Run one validation-shaped pass as well.
- [ ] Synchronize CUDA before exit, then report elapsed prime time, cache directory, compiled signatures, graph breaks/recompiles, and cache hit/miss information available from PyTorch.
- [ ] Avoid checkpoints, experiment logs, and result claims during priming. This command prepares compiler artifacts only.
- [ ] Run the prime command in two fresh processes and confirm that the second process reuses the local cache.
- [ ] Document that cache entries are local executable build artifacts. Do not commit them or load cache bundles from untrusted sources.

### 4. Remove known graph-break hazards

- [ ] Replace the `loss.item()` fallback branch with configuration-time logic so the loss path does not synchronize the GPU or branch on tensor data.
- [ ] Audit training-time state reset/detach mutations and isolate them from compiled regions when necessary.
- [ ] Inspect graph breaks and recompiles with PyTorch compiler diagnostics on one representative run per architecture.
- [ ] Do not depend on private Dynamo flags merely to force recurrent layers into a graph; document partial/eager execution when that is the supported behavior.

### 5. Validate correctness and persistence

- [ ] Add focused tests for config/CLI behavior and for invoking the selected compile boundary.
- [ ] Compare eager and compiled forward/loss results within documented floating-point tolerances.
- [ ] Run short training and validation smoke workloads through Lightning on CPU and CUDA where available.
- [ ] Save a checkpoint from compiled training, load it into a fresh eager model, resume training, and confirm the state-dict schema is unchanged.
- [ ] Confirm deterministic-mode behavior is either preserved or its limitation is clearly reported.

### 6. Validate export after compiled training

- [ ] Export a checkpoint produced by compiled training through the normal eager reconstruction path.
- [ ] Validate ONNX Runtime parity for the currently supported LSTM, GRU, and WaveNet exports, including dynamic batch/time shapes where supported.
- [ ] Validate TorchScript for S4D and the other supported model families.
- [ ] Preserve the existing export support matrix: this issue does not add S4D ONNX or stateful real-time ONNX support.
- [ ] Coordinate with #37 rather than introducing a second ONNX workaround.

### 7. Run real end-to-end benchmarks and decide support per architecture

- [ ] Run at least one real-audio training epoch in eager and compiled modes for each architecture on the same CUDA device.
- [ ] Run at least three fresh-process warmed measurements for each viable compiled configuration.
- [ ] Measure compiled cold start with an isolated empty cache, then measure normal startup and training after running the cache-prime command. Do not mix cold-cache and warm-cache timings.
- [ ] Report median and variability for compile latency, warmed training time/throughput, peak CUDA memory, graph breaks/recompiles, and validation loss.
- [ ] Treat a mode as beneficial only when its warmed result is repeatably faster without a correctness, memory, checkpoint, resume, or export regression.
- [ ] Publish a per-architecture decision: recommended mode, supported but neutral, or unsupported/not worthwhile.
- [ ] Keep eager as the default unless the measured result supports a separate, reviewed default change.

### 8. Document the operational contract

- [ ] Document how to enable compilation, expected first-run overhead, cache/recompilation behavior, diagnostics, and common failure messages.
- [ ] Store the machine-readable benchmark results and a generated Markdown summary.
- [ ] Update example configs only for architectures where the benchmark supports a recommendation.

## Parallel work after step 2

Once the config and compile boundary are fixed, these tracks can proceed independently:

1. WaveNet/S4D compiler diagnostics and performance experiments.
2. LSTM/GRU compiler diagnostics and performance experiments.
3. Checkpoint/resume and ONNX/TorchScript compatibility validation.
4. Persistent cache tooling, benchmark reporting, documentation, and focused automated tests.

The final support decision and documentation depend on all four tracks.

## Acceptance criteria

- [ ] `scripts/train.py` can run the same configuration eagerly or with explicitly requested compilation.
- [ ] Compiled training defaults to a persistent local cache outside the repository, prints the selected path, and honors explicit environment/config overrides.
- [ ] The cache-prime command populates forward, backward, and validation compiler artifacts for the exact training config without reading the full dataset or writing a checkpoint.
- [ ] A second fresh-process prime run demonstrates cache reuse, and the first full training run can use the same entries.
- [ ] Eager behavior remains the default and an explicitly requested compile failure is not hidden.
- [ ] Compiled training completes a real paired-audio Lightning workload on CUDA for every architecture classified as supported.
- [ ] Published results separate compile startup cost from warmed performance and include at least three fresh-process measurements.
- [ ] Supported compiled modes preserve loss/output behavior within tolerance and show no checkpoint, resume, validation, or memory regression.
- [ ] A checkpoint from compiled training loads into the ordinary eager model with the existing key schema.
- [ ] Post-training exports pass the repository's current ONNX/TorchScript runtime parity checks and do not expand or reduce the documented support matrix.
- [ ] Graph breaks/recompiles are measured and summarized for each architecture; unsupported or non-beneficial cases are documented rather than forced.
- [ ] Existing tests continue to pass.

## Non-goals

- Compiling real-time inference or committing compiler caches.
- Adding Redis-backed remote caching or distributing cache bundles between machines.
- Changing precision, TF32, data loading, optimizer, or batch/segment size in the same benchmark comparison.
- Forcing `torch.compile` on every model or making it the global default without measured evidence.
- Expanding ONNX support beyond the work tracked in #37.

## Related work

- #14 provides inference/export benchmarking and artifact paths.
- #15 provides the fixed evaluation workload and model-quality context.
- #37 owns PyTorch 2.11 ONNX compatibility.
- PyTorch's compiler caching documentation defines the local FX graph, AOTAutograd, Inductor, Triton, and autotuning caches: https://docs.pytorch.org/tutorials/recipes/torch_compile_caching_configuration_tutorial.html
